packages/require/lib/index.ts

```diff
@@ -0,0 +1,130 @@
+import path from "path";
+import TruffleConfig from "@truffle/config";
+import { options as expectOptions } from "@truffle/expect";
+import { Web3Shim, createInterfaceAdapter } from "@truffle/interface-adapter";
+
+import vm from "vm";
+import Module from "module";
+
+import originalRequire from "original-require";
+
+import { ExecOptions, RequireOptions } from "./types";
+import { compile, registerTsNode } from "./compiler";
+
+export function file(options: RequireOptions) {
+  const scriptPath = path.resolve(options.file);
+
+  expectOptions(options, ["file"]);
+
+  const conf = TruffleConfig.default().with(options);
+
+  const scriptModule = new Module(scriptPath);
+
+  // Provide all the globals listed here: https://nodejs.org/api/globals.html
+  const sandbox: { [index: string]: any } = {
+    __filename: path.basename(scriptPath),
+    __dirname: path.dirname(scriptPath),
+    Buffer,
+    clearImmediate,
+    clearInterval,
+    clearTimeout,
+    console,
+    exports: scriptModule.exports,
+    global,
+    process,
+    setImmediate,
+    setInterval,
+    setTimeout,
+    module: scriptModule,
+    config: conf,
+    artifacts: conf.resolver,
+    require: (id: string) => {
```

Comment:
This implementation seems fine, probably.

I'd love an explanation as to why it is necessary in the first place. Why can't we pass in `require` (i mean `original-require`, once webpacked) directly?

At first I thought the answer was because the `require` available here shares a global scope with the parent script, and that could be a problem, but this doesn't seem to change that.

Is this all because of relative path issues? If so, I think there are built-in methods to resolve those we could use instead of all this. https://nodejs.org/docs/latest/api/modules.html#requiremain

Reply:
It looks like this function is here primarily to handle node modules that are installed in the same directory as the script?

Reply:
> I'd love an explanation as to why it is necessary in the first place.

This code is just ported over from the JS version, and it's 5 years old, so I don't have all of the context.

My understanding is that it's basically attempting to resolve webpacked modules first, then it falls back to the typical node built-in require and attempts to resolve them externally as per usual node behavior.

However as you sort of alluded to, it's a somewhat poor approximation of the actual `require`. I did attempt to rewrite it at one stage (with the goal of executing modules required by scripts running in the child context within that same child context) but ran into some complexities around relative paths for transitive dependencies.

The `vm2` module has solved this to some extent, but then they have hardened their VM sandbox to prevent modification to builtins (`source-map-support` apparently modifies built-ins), and annoyingly they always force strict mode, even if your source doesn't include the `"use strict";` directive.